### PR TITLE
[crypto/ec] for ECC parameters with NULL or zero cofactor, compute it (1.1.0)

### PR DESCRIPTION
**Pull request by bbbrumley**
Backport of #9727 to 1.1.0. The test there isn't at all compatible with 1.1.0. So to test this, I did the following:

### hack test

Ignored the passed cofactor, always computed it, then added an assert:

```
diff --git a/crypto/ec/ec_lib.c b/crypto/ec/ec_lib.c
index eaf44ccef9..af55de37d1 100644
--- a/crypto/ec/ec_lib.c
+++ b/crypto/ec/ec_lib.c
@@ -366,7 +366,7 @@ int EC_GROUP_set_generator(EC_GROUP *group, const EC_POINT *generator,
         return 0;
 
     /* Either take the provided positive cofactor, or try to compute it */
-    if (cofactor != NULL && !BN_is_zero(cofactor)) {
+    if (0 && cofactor != NULL && !BN_is_zero(cofactor)) {
         if (!BN_copy(group->cofactor, cofactor))
             return 0;
     } else if (!ec_guess_cofactor(group)) {
@@ -374,6 +374,9 @@ int EC_GROUP_set_generator(EC_GROUP *group, const EC_POINT *generator,
         return 0;
     }
 
+#include <assert.h>
+    assert(BN_cmp(group->cofactor, cofactor) == 0);
+
     /*
      * Some groups have an order with
      * factors of two, which makes the Montgomery setup fail.
```
Then `make test` went swell.

### EVP test

I added two explicit parameter PEMs with no cofactor to the EVP tests. The two key pairs are from [RFC 6979](https://tools.ietf.org/html/rfc6979) for P-256 and P-384. Unfortunately my team's tooling only currently supports NIST curves over prime fields.

### gdb

I checked the codepath in `gdb` to make sure it's taking the proper early exit to the ladder.

### CODE REVIEW

File: CHANGES [hunk 1]
```diff
@@ -16,6 +16,11 @@
      (CVE-2019-1552)
      [Richard Levitte]
 
+  *) Compute ECC cofactors if not provided during EC_GROUP construction. Before
+     this change, EC_GROUP_set_generator would accept order and/or cofactor as
+     NULL. After this change, only the cofactor parameter can be NULL. It also
+     does some minimal sanity checks on the passed order.
+     [Billy Bob Brumley]
```
Comment (romen) [inline CHANGES]: nit: The CHANGES entry should go to the top of the top section.

Comment (bbbrumley) [reply CHANGES]: Fixed in 2f808454c6aee59e78365cc42c3135b6fae0c356

File: CHANGES [hunk 2]
```diff
@@ -9,14 +9,19 @@
 
  Changes between 1.1.0k and 1.1.0l [xx XXX xxxx]
 
+  *) Compute ECC cofactors if not provided during EC_GROUP construction. Before
+     this change, EC_GROUP_set_generator would accept order and/or cofactor as
+     NULL. After this change, only the cofactor parameter can be NULL. It also
+     does some minimal sanity checks on the passed order.
```
Comment (mattcaswell) [inline CHANGES]: We should include the CVE number in the CHANGES entry (like the entry below). I guess we missed this from the 1.1.1. version.

File: include/openssl/ec.h [hunk 3]
```diff
@@ -1568,6 +1568,7 @@ int ERR_load_EC_strings(void);
 # define EC_R_SLOT_FULL                                   108
 # define EC_R_UNDEFINED_GENERATOR                         113
 # define EC_R_UNDEFINED_ORDER                             128
+# define EC_R_UNKNOWN_COFACTOR                            136
```
Comment (mattcaswell) [inline include/openssl/ec.h]: This should be 164 to maintain ABI compatibility between 1.1.0 and 1.1.1

Comment (romen) [reply include/openssl/ec.h]: Fixed by 9b29942

File: CHANGES (hunk 2 above)
Comment (romen) [reply CHANGES]: Fixed by dbbdea6d45b177167fc440d9a96d85579f015ed3

**Comment by romen** [comment]
I know @bbbrumley is busy until Sunday night, and I am authorized to apply changes to his PR for this level of triviality, so in the interest of expediting the review process in view of the upcoming release day, I took the liberty of pushing the required changes.

**Comment by romen** [comment]
I have also personally confirmed the manual "hack testing" done by @bbbrumley 

I will now proceed to merge

**Comment by romen** [comment]
Merged to `1.1.0-stable` with:

- c31be97c64 [test/recipes/30-test_evp_data] computing ECC cofactors: regression test                                                                                                      
- 7c1709c2da [crypto/ec] for ECC parameters with NULL or zero cofactor, compute it

**Comment by richsalz** [comment]
Why does the CHANGES addition have to go at the top?  Is chronology within a release important?  After all, many more CHANGES entries are going to appear before the official release; does it matter if an entry is the 32nd or the 45th?

**Comment by mspncp** [comment]
> Why does the CHANGES addition have to go at the top?

I don't understand what disturbs you; isn't it common practice to add new CHANGES entries at the top?

**Comment by richsalz** [comment]
Every time I rebase, I have to fix a merge conflict to put my entry back on the top because there's a new top entry and the context-diff gets confused.  If I just let it slide down, then future CHANGES rebasing will not cause a problem.

**Comment by mspncp** [comment]
Oh, _now_ I understand your point. Nice trick. Still I prefer that new entries appear at the top in the order in which the pull requests were merged.

Also, your nice trick has a little flaw: letting your entry sink down when rebasing only guarantuees a conflict-free experience if nobody else uses the same strategy. Because otherwise new entries on master will also appear at lower places and conflict with your entry.

So an improved strategy would be: while rebasing a pull request on master, let the entry sink down, but when the pull request gets finally merged, the entry should be moved back to the top. Unfortunately, only committers could do this last step.

**Comment by mspncp** [comment]
_(I wrote 'master', but of course the same reasoning applies to the stable branches)_

**Comment by richsalz** [comment]
So your saying it only works for me?  That's fine :)

**Comment by mspncp** [comment]
Sorry, no. It works only until you tell other people. Which you just did. 😆

**Pull request closed without merging**